### PR TITLE
[Vouchers] Flat rate of any amount

### LINKED ISSUE #10727: [Vouchers] Include unfixed amounts and percentage rates
## Description

Extending the first bare minimum iteration of vouchers this next stage will expand the functionality to include:
- Both Percentage and Flat Amount for Voucher Type
- The Amount can be set to anything. For percentage this is between 0 and 100 as a double. For Flat Amount this is a non-negative double. 
Screenshot from 2023-04-20 07-32-19
_Screenshot from designs [here](https://www.figma.com/file/ZOe1Xmz4kg9wnHw1XFu389/Voucher-Back-office?node-id=47-301)_

The rate will be reflected in the list page and will be calculated accurately at checkout.

## Acceptance Criteria & Tests

1. Create voucher codes in the backoffice - using a range of percentages to test edge cases (0, 100 etc) and amounts (more than order total, exactly order total etc)
2. Apply to a checkout flow and see the discount applied
3. If the order total goes below zero when the code is applied it should stay at zero
4. Disregard (at this stage) any case that additional negative fees apply to the order
5. Voucher codes should be able to be used unlimited number of times
6. Voucher codes should apply to any shoppers
7. Voucher codes should not apply in enterprises that have not enabled them. If the enterprise has enabled no voucher code the voucher code box should not be visible at checkout
8. If the voucher code covers the cost of the whole order, just display the full range of payment methods, no clever logic at this stage.
10. Translations as standard
11. Unit tests as standard

### PR DESCRIPTION
**Pull request by rioug**
#### What? Why?

- First part of  #10727

Add the ability to have a flat rate voucher of any amount

#### What should we test?

* Enable vouchers feature
* Visit admin/feature-toggle/features and enable or add vouchers feature

In the backoffice logged as an enterprise user:

- Visit the Enterprise tab and click on "Settings" for a shop of your choosing, and click on the "vouchers" in left hand side menu
- Click on "Add New button" --> New Voucher page should now have an "Amount" input
- Enter a voucher code and an amount and click "Save"
    --> It should redirect you to the vouchers page on the enterprise edit page
    --> You should now see a list of vouchers with the one you just created, included the entered amount

On the website logged as a customer :

- Add one or more product in your cart from the shop above
- Proceed to checkout
- Enter details and shipping info, click on "Next - Payment method"
- Enter the voucher code and click "Apply"
--> You should see a "$xx Voucher" box where xx is the amount you entered when creating a voucher
- choose a payment method and click on "Next - Order summary"
--> you should see the voucher applied to the order in the order summary column, with the correct amount
- Click on "Complete order"
--> you should see the voucher applied to the order in the order summary with the correct amount 

#### Release notes

Changelog Category: User facing changes

The title of the pull request will be included in the release notes.

### CODE REVIEW

File: app/models/spree/adjustment.rb [hunk 1]
```diff
@@ -44,6 +44,8 @@ class Adjustment < ApplicationRecord
 
     belongs_to :tax_rate, -> { where spree_adjustments: { originator_type: 'Spree::TaxRate' } },
                foreign_key: 'originator_id'
+    belongs_to :voucher, -> { where spree_adjustments: { originator_type: 'Spree::Voucher' } },
```
Comment (Matt-Yorkley) [inline app/models/spree/adjustment.rb:47]:
```suggestion
    belongs_to :voucher, -> { where spree_adjustments: { originator_type: 'Voucher' } },
```
This class isn't in the `Spree` namespace, right?

Comment (Matt-Yorkley) [reply app/models/spree/adjustment.rb:47]: Ah I see this builds on the previous PR, I'll move my review there... :+1:

File: spec/factories/voucher_factory.rb [hunk 2]
```diff
@@ -0,0 +1,8 @@
+# frozen_string_literal: true
+
+FactoryBot.define do
+  factory :voucher do
+    enterprise { create(:distributor_enterprise) }
+    amount { rand(200.0) }
+  end
```
Comment (mkllnk) [inline spec/factories/voucher_factory.rb:7]: Great to have a factory. When I tried it out I noticed that enterprise is not a required field. We still have an old Rails config:

https://github.com/openfoodfoundation/openfoodnetwork/blob/482d438ec93f91bcc4f6aafa307fdfa2cffeb4d9/config/application.rb#L229

So we need to add `optional: false` to the model, I think.

Also, when you just `build(:voucher)` then it *creates* an enterprise unnecessarily. It would be better to use `build` here because it's much faster when you don't need to persist.

I personally don't like random test data. It can make tests flaky and they are more difficult to debug when they are not deterministic. Sometimes randomness can reveal bugs and therefore it's a trade-off but I prefer deterministic tests.

**Review by mkllnk — CHANGES_REQUESTED** [review]
Nice work!

**Comment by rioug** [comment]
New commit start here : https://github.com/openfoodfoundation/openfoodnetwork/pull/10761/commits/2da76b203653704a8146f809c10c4491571d948c

File: db/schema.rb [hunk 3]
```diff
@@ -1189,26 +1189,27 @@
...
   end
 
-  create_table "webhook_endpoints", force: :cascade do |t|
-    t.string "url", null: false
-    t.datetime "created_at", null: false
-    t.datetime "updated_at", null: false
-    t.bigint "user_id", default: 0, null: false
-    t.index ["user_id"], name: "index_webhook_endpoints_on_user_id"
-  end
-
```
[the remark this answers is not in the document]
Comment (rioug) [reply db/schema.rb]: I reset my DB but I still have some weird difference with `schema.rb` from master ?? Anyway I fixed my commit to only show the new changes

**Review by mkllnk — APPROVED** [review]
Great. Thank you for addressing it all.

I noticed that I get weird results in db/schema.rb on master as well. I think that something must have gotten wrong when merging. I'll open a PR to correct master.

File: app/views/split_checkout/_voucher_section.cable_ready.haml [hunk 4]
```diff
@@ -10,10 +10,10 @@
...
           - # This might not be true, ie payment method including a fee which wouldn't be covered by voucher or tax implication raising total to be bigger than the voucher amount ?
-          - if voucher_adjustment.originator.value > order.total
+          - if voucher_adjustment.originator.amount > order.total
             .checkout-input
               %span.formError.standalone
                 = t("split_checkout.step2.voucher.warning_forfeit_remaining_amount")
         - else
           = text_field_tag "[order][voucher_code]", params.dig(:order, :voucher_code), data: { action: "input->toggle-button-disabled#inputIsChanged", }, placeholder: t("split_checkout.step2.voucher.placeholder") , class: "voucher"
-          = submit_tag t("split_checkout.step2.voucher.apply"), name: "apply_voucher", disabled: true, class: "button cancel voucher", "data-disable-with": false, data: { "toggle-button-disabled-target": "button" }
+          = submit_tag t("split_checkout.step2.voucher.apply"), name: "apply_voucher", disabled: true, class: "button cancel voucher", "data-disable-with": false, data: { "toggle-button-disabled-target": "button" }
```
Comment (dacook) [inline app/views/split_checkout/_voucher_section.cable_ready.haml:19]: Oops, a couple of spaces were added at end of line here.

File: db/migrate/20230418030646_add_amount_to_vouchers.rb [hunk 5]
```diff
@@ -0,0 +1,5 @@
+class AddAmountToVouchers < ActiveRecord::Migration[7.0]
+  def change
+    add_column :vouchers, :amount, :decimal, precision: 10, scale: 2, null: false, default: 0
```
[the remark this answers is not in the document]
Comment (dacook) [reply db/migrate/20230418030646_add_amount_to_vouchers.rb:3]: Is a default necessary? If an amount isn't provided, then we want an error to be thrown don't we?

File: app/controllers/admin/vouchers_controller.rb [hunk 6]
```diff
@@ -26,7 +26,10 @@ def create
     private
 
     def load_enterprise
-      @enterprise = Enterprise.find_by(permalink: params[:enterprise_id])
+      @enterprise = OpenFoodNetwork::Permissions
+        .new(spree_current_user)
+        .editable_enterprises
+        .find_by(permalink: params[:enterprise_id])
```
Comment (dacook) [inline app/controllers/admin/vouchers_controller.rb:32]: Side comment..
It looks funny to see permalink coming from  `params[:enterprise_id]`. I can see this is common across the whole admin interface.. but I have to admit I can't see why!

File: spec/system/admin/vouchers_spec.rb [hunk 7]
```diff
@@ -11,6 +11,7 @@
 
   let(:enterprise) { create(:supplier_enterprise, name: 'Feedme') }
   let(:voucher_code) { 'awesomevoucher' }
+  let(:amount) { 25 }
```
Comment (dacook) [inline spec/system/admin/vouchers_spec.rb:14]: `±` I think we also need a spec here for updating an existing voucher

**Review by dacook — COMMENTED** [review]
I think this is fine as-is, but have a few comments to potentially improve.

File: spec/system/admin/vouchers_spec.rb (hunk 7 above)
Comment (rioug) [reply spec/system/admin/vouchers_spec.rb:14]: There is currently no way to update a voucher, do you still think we need a spec for it ?

Comment (dacook) [reply spec/system/admin/vouchers_spec.rb:14]: Oh, sorry I was mistaken!

File: db/migrate/20230418030646_add_amount_to_vouchers.rb (hunk 5 above)
[the remark this answers is not in the document]
Comment (rioug) [reply db/migrate/20230418030646_add_amount_to_vouchers.rb:3]: I think you are right, my thinking was a voucher with an amount 0 is probably not a big problem, the worse than can happen is you apply a voucher that won't change the order amount..

**Comment by rioug** [comment]
This will need to be rebased on https://github.com/rioug/openfoodnetwork/tree/10432-vouchers-bare-minimum-checkout (or master if 10432-vouchers-bare-minimum-checkout has been merged)

**Review by mkllnk — APPROVED** [review]

File: spec/factories/voucher_factory.rb (hunk 2 above)
Comment (dacook) [reply spec/factories/voucher_factory.rb]: @mkllnk I can see that the belongs_to required issue is resolved now, but I also think it's worth removing the random value. 

Although it could potentially help uncover new cases it would be random and suck up a lot of tech debt time with perhaps no benefit. So I don't think we should set this precedent, do you agree if I change this to a value like $15?

**Review by dacook — CHANGES_REQUESTED** [review]
I think we should still change the random value in the factory.

**Review by dacook — CHANGES_REQUESTED** [review]
I think we should still change the random value in the factory.

**Review by dacook — CHANGES_REQUESTED** [review]
I think we should still change the random value in the factory.

**Comment by mkllnk** [comment]
Yes, I would be happy with a fixed value.

**Review by dacook — APPROVED** [review]

**Comment by drummer83** [comment]
Hi @rioug and @mkllnk,
So exciting to play with this new feature! Since it was my first time I have also noted some things which we may want to look into at some point, but they are out of scope of this PR.

### General remark after staging
- Previously generated vouchers are set to Amount = 0 (instead of 10 $ when created). I guess we can live with this, since it is not in production yet.

### Positive test cases
- only numbers > 0 are accepted :heavy_check_mark: 
- the new voucher is added to the vouchers list :heavy_check_mark: 
- checkout using the voucher works as expected :heavy_check_mark: 
- order confirmation page shows the voucher :heavy_check_mark: 
- confirmation email shows the voucher :heavy_check_mark: 
- vouchers can be applied unlimited times :heavy_check_mark: 
- vouchers can be applied by any customer :heavy_check_mark: 
- vouchers are not displayed when checking out at an enterprise which has no voucher codes :heavy_check_mark: 

### Proposed adjustments of backoffice design
- Main page
  - Shouldn't be available for hub profile and shop profile enterprises (sells = none)
  - 'Add new' button is currently above the table --> Comparing to Payment/Shipping methods: Name it 'Create new voucher +' and place it below the table
image
- Create new voucher page
  - Use header design from Payment/Shipping methods: 'New Voucher' & button 'Back to voucher list'
  - Include the 'Cancel' and 'Create' buttons in the footer of the page (see Payment/Shipping methods)
image
  - Is a multi-line text field required for the voucher code? --> It's not possible to use a voucher with two or more lines at checkout (Error: voucher not found).
- Default value should be 0.00 (instead of 0.0 which is untypical for money amounts).
- Entering a comma as decimal separator shows flash error - we have just recently decided to go for more browser validation
- Entering letters shows the flash error - we have just recently decided to go for more browser validation

### Proposed adjustments for backoffice functionality
- Edit vouchers
- Delete vouchers

### Proposed adjustments for checkout
- Irritating error message when voucher can't be found and no red highlighting of the voucher field.
image

### Negative test cases
- if after applying the voucher you go back to edit your basket and the order total is 0 $ after updating the cart, you then click checkout, enter your details and then click 'next - payment details', then you are redirected to the main page, but the order is placed (email confirmation is sent)

https://github.com/openfoodfoundation/openfoodnetwork/assets/1790176/80b84033-30d8-4834-8a1a-b205c724cdb1

- the voucher is displayed on the /cart page after applying it on step 2 of checkout. However it's value is not being calculated into the order total (:arrow_right: #10855):

https://github.com/openfoodfoundation/openfoodnetwork/assets/1790176/0b8c030e-4036-448f-a720-14b39019b3b1

- if a voucher code is entered but not yet applied and no payment method is selected when clicking 'next', then the 'apply' button remains disabled. You have to change the value in the voucher code field before it can be applied.

https://github.com/openfoodfoundation/openfoodnetwork/assets/1790176/9bf50a55-6b86-4d02-8d96-998bdbc5af3e

## Result

The requested feature has been implemented according to specification. We can merge this one here. I will open issues for the negative test cases and some of the improvements.

Thanks! :pray:

**Comment by mkllnk** [comment]
Thank you, excellent testing. Definitely some edge cases to work out before it can be released.
